packages/go/analysis/ad/esc3.go

```diff
@@ -142,60 +142,71 @@ func PostADCSESC3(ctx context.Context, tx graph.Transaction, outC chan<- analysi
 	return nil
 }
 
-func PostEnrollOnBehalfOf(certTemplates []*graph.Node, operation analysis.StatTrackedOperation[analysis.CreatePostRelationshipJob]) error {
+func PostEnrollOnBehalfOf(domains, enterpriseCertAuthorities, certTemplates []*graph.Node, cache ADCSCache, operation analysis.StatTrackedOperation[analysis.CreatePostRelationshipJob]) error {
 	versionOneTemplates := make([]*graph.Node, 0)
 	versionTwoTemplates := make([]*graph.Node, 0)
-
 	for _, node := range certTemplates {
 		if version, err := node.Properties.Get(ad.SchemaVersion.String()).Float64(); errors.Is(err, graph.ErrPropertyNotFound) {
 			log.Warnf("Did not get schema version for cert template %d: %v", node.ID, err)
 		} else if err != nil {
 			log.Errorf("Error getting schema version for cert template %d: %v", node.ID, err)
+		} else if version == 1 {
+			versionOneTemplates = append(versionOneTemplates, node)
+		} else if version >= 2 {
+			versionTwoTemplates = append(versionTwoTemplates, node)
 		} else {
-			if version == 1 {
-				versionOneTemplates = append(versionOneTemplates, node)
-			} else if version >= 2 {
-				versionTwoTemplates = append(versionTwoTemplates, node)
-			} else {
-				log.Warnf("Got cert template %d with an invalid version %d", node.ID, version)
-			}
+			log.Warnf("Got cert template %d with an invalid version %d", node.ID, version)
 		}
 	}
 
-	operation.Operation.SubmitReader(func(ctx context.Context, tx graph.Transaction, outC chan<- analysis.CreatePostRelationshipJob) error {
-		if results, err := EnrollOnBehalfOfVersionTwo(tx, versionTwoTemplates, certTemplates); err != nil {
-			return err
-		} else {
-			for _, result := range results {
-				if !channels.Submit(ctx, outC, result) {
-					return nil
-				}
-			}
+	for _, domain := range domains {
```

Comment:
Might there be a way to combine this EnrollOnBehalfOf processing into the ESC3? I ask in case there is a way to avoid doing this double loop through domains and eca's twice (here and in the PostADCS func).

Reply:
I like that idea! I think we could throw the EnrollOnBehalfOf logic under ESC3 and put the nodes that should get EnrollOnBehalfOf edge into variables, so we can perform lookups in those variables instead of checking for EnrollOnBehalfOf edges. At the end, we could create both the ESC3 and EnrollOnBehalfOf edges. That said, I don't have time to implement and test that before sometimes next year..

Reply:
Makes sense. This would take some time to clean up!